### PR TITLE
Content Caching

### PR DESCRIPTION
**Pull request by gregpriday**
The main goal of this content caching isn't so much speed as it is making Page Builder content appear more like standard WordPress content to other plugins. We could also enhance this to use WordPress object caching in the future, which would open it up to things like Memcache.

### CODE REVIEW

**Comment by gregpriday** [comment]
Just make sure you have the latest develop of Widgets Bundle to get https://github.com/siteorigin/so-widgets-bundle/commit/1c06048787e4b6f51c6d969b369f778e7e451663

File: inc/admin.php [hunk 1]
```diff
@@ -174,33 +181,29 @@ function save_post( $post_id, $post ) {
...
-				$post->post_content = $post_content;
-
-				// Update the post_content for this post
+				// Store a version of the HTML in post_content
+				SiteOrigin_Panels_Post_Content_Filters::add_filters();
+				$GLOBALS[ 'SITEORIGIN_PANELS_POST_CONTENT_RENDER' ] = true;
+				$post_content = SiteOrigin_Panels_Renderer::single()->render( $post_id, false, $panels_data );
+				$post_css = SiteOrigin_Panels_Renderer::single()->generate_css( $post_id, $panels_data );
+				SiteOrigin_Panels_Post_Content_Filters::remove_filters();
+				$GLOBALS[ 'SITEORIGIN_PANELS_POST_CONTENT_RENDER' ] = false;
```
Comment (gregpriday) [inline inc/admin.php]: Aaah yeah, nicely spotted. Will change this to unset.

File: inc/cache.php [hunk 2]
```diff
@@ -0,0 +1,155 @@
+<?php
+
+class SiteOrigin_Panels_Cache {
```
Comment (gregpriday) [inline inc/cache.php]: Yeah, agreed. Have renamed it to `SiteOrigin_Panels_Cache_Renderer`. I'll look a little bit more at making things more obvious.

File: inc/cache.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+	private $cache_render;
+	private $post_id;
```
Comment (gregpriday) [inline inc/cache.php]: True. Have removed all $this->post_id stuff.

File: siteorigin-panels.php [hunk 4]
```diff
@@ -56,6 +53,21 @@ function __construct() {
 		if( siteorigin_panels_setting( 'bundled-widgets' ) ) {
 			require_once plugin_dir_path( __FILE__ ) . 'widgets/widgets.php';
 		}
+
+		SiteOrigin_Panels_Widget_Shortcode::init();
```
Comment (gregpriday) [inline siteorigin-panels.php]: We definitely are. I often just make something a singleton out of habit, without actually thinking if it's singleton worthy or not. In a refactor, we should look at un-singling some singletons.

File: inc/widget-shortcode.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	 * @param $attr
+	 * @param $content
+	 *
+	 * @return string
+	 */
+	static function shortcode( $attr, $content ){
+		$attr = shortcode_atts( array(
+			'class' => false,
+			'id' => '',
+		), $attr, 'panels_widget' );
```
Comment (gregpriday) [inline inc/widget-shortcode.php]: Aaah yeah. Was originally going to have different mechanisms for Widgets Bundle and Page Builder. In the end I combined them.

File: inc/widget-shortcode.php [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+			in_array( get_class( $widget ), self::$text_widgets )
+		) {
+			return $html;
+		}
+
+		return self::get_shortcode( $widget, $args, $instance );
+	}
+
+	static function encode_data( $data ){
+		return '<input type="hidden" value="' . htmlentities( json_encode( $data ), ENT_QUOTES ) . '" />';
```
Comment (gregpriday) [inline inc/widget-shortcode.php]: Not sure what you mean here. Can you clarify?

File: inc/widget-shortcode.php [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	}
+
+	static function decode_data( $string ){
+		preg_match( '/value="(.*)" \/>/', trim( $string ), $matches );
```
Comment (gregpriday) [inline inc/widget-shortcode.php]: I got some weird results without that, so just added it in. I'll take another look. I think I wasn't encoding something properly, though, which I fixed. So can probably remove that last part.

Maybe matching `value="([^"]*)"` would be better.

**Comment by gregpriday** [comment]
@braamgenis: there was an issue in develop with adding styling. It triggered a warning for a missing cell_index. This might have been the warnings you saw.

File: inc/widget-shortcode.php (hunk 6 above)
Comment (braamgenis) [inline inc/widget-shortcode.php]: Haha! Disregard. Tired brain is tired. :sweat_smile:

**Pull request merged**